src/OpenTelemetry.Exporter.Zipkin/ZipkinExporter.cs

```diff
@@ -89,6 +90,14 @@ public override ExportResult Export(in Batch<Activity> batch)
             }
             catch (Exception ex)
             {
+                if (!content.BatchFlushed)
+                {
+                    foreach (Activity activity in batch)
```

Comment:
Wasn't sure the best approach here. Thought about adding a `Drain` on `Batch<T>` but I didn't want to increase the API surface. This is really a problem that will impact all exporters. They _must_ drain the batch, or it will fill up. Maybe this should be the responsibility of `BatchExportProcessor<T>`?

/cc @alanwest @cijothomas @reyang

Reply:
Something like catching the exception in https://github.com/open-telemetry/opentelemetry-dotnet/blob/05bba18f4d5748571dc986dcdce38580f6083c6b/src/OpenTelemetry/BatchExportProcessor.cs#L220, and drain the batch if there are items left (not consumed by the exporter)?

Reply:
That's what I was thinking. BUT... I don't think we want Export to throw? It does return a success/failure status. We could generalize it something like... before the call we set to locals how many items are being exported (head & tail of the circular buffer). Then after the call we drain anything that is remaining, for whatever reason.

Reply:
Sounds very reasonable to me 👍

Reply:
Could this be one possible approach:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/05bba18f4d5748571dc986dcdce38580f6083c6b/src/OpenTelemetry/Batch.cs#L92

```csharp
public void Dispose()
{
    while (this.MoveNext())
    {
        // drain
    }
}
```

Reply:
The problem I see there is it's at the Enumerator level. So if the Exporter got to the point of creating the enumeration, we would be OK. But if it didn't get that far, which is the case here with Zipkin having a connection failure, the Enumerator was never created so it couldn't drain itself.

Reply:
Batch<T>.Dispose() 😂

Reply:
I think I might be able to make that work. I'll give it a shot.

Reply:
@reyang Updated. While I was in `Batch<T>` I also made it resilient to people trying to enumerate the same batch multiple times.